plugins/inputs/t128_graphql/response_processor.go

```diff
@@ -13,56 +13,287 @@ type ProcessedResponse struct {
-		//TODO: move to init if keeping - MON-315
-		processedResponse := &ProcessedResponse{Fields: map[string]interface{}{}, Tags: map[string]string{}}
-
-		for fieldRenamed, fieldName := range fields {
-			data := node[fieldName]
-
-			if strings.Contains(fieldName, "/") {
-				//TODO: clean this up, it's the second time JSONPointer() is called on the response - MON-315
-				nestedObj, err := child.JSONPointer("/" + fieldName)
-				if err != nil {
-					return nil, fmt.Errorf("unexpected response for collector %s: field %s", collectorName, fieldName)
-				}
-				data = nestedObj.Data()
+/*
+ProcessResponse takes in a query response, pulls out the desired data and stores it in a struct. To find the data,
+ProcessResponse traverses the response tree recursively and appropriately merges the data returned by each child.
+
+Args:
+	jsonData example - see example below
+	fields example - map[string]string{
+			"/data/allRouters/nodes/peers/nodes/paths/adjacentAddress": "adjacent-address",
+			"/data/allRouters/nodes/peers/nodes/paths/uptime": "uptime"
+		}
+	tags example - map[string]string{
+			"/data/allRouters/nodes/peers/nodes/name": "peer-name",
+		}
+
+Example:
+	For the the input given above along with the following jsonData input
+
+	{"data":
+		{
+			"allRouters": {
+			"nodes": [
+				{
+				"peers": {
+					"nodes": [
+					{
+						"name": "AZDCBBP1",
+						"paths": [
+							{
+								"uptime": 188333176,
+								"adjacentAddress": "12.51.52.30"
+							},
+							{
+								"uptime": 82247253,
+								"adjacentAddress": "12.51.52.30"
+							}
+						]
+					},
+					{
+						"name": "AZDCLTEP1",
+						"paths": [
+							{
+								"uptime": 162241794,
+								"adjacentAddress": "12.51.52.22"
+							},
+							{
+								"uptime": 82247352,
+								"adjacentAddress": "12.51.52.22"
+							}
+						]
+					}]
+				}}
+			]
+		}
+	}
+
+	ProcessResponse() will produce the following output
+
+	[]*plugin.ProcessedResponse{
+		&plugin.ProcessedResponse{
+			Fields: map[string]interface{}{"adjacent-address": "12.51.52.30", "uptime": 188333176.0},
+			Tags:   map[string]string{"peer-name": "AZDCBBP1"},
+		},
+		&plugin.ProcessedResponse{
+			Fields: map[string]interface{}{"adjacent-address": "12.51.52.30", "uptime": 82247253.0},
+			Tags:   map[string]string{"peer-name": "AZDCBBP1"},
+		},
+		&plugin.ProcessedResponse{
+			Fields: map[string]interface{}{"adjacent-address": "12.51.52.22", "uptime": 162241794.0},
+			Tags:   map[string]string{"peer-name": "AZDCLTEP1"},
+		},
+		&plugin.ProcessedResponse{
+			Fields: map[string]interface{}{"adjacent-address": "12.51.52.22", "uptime": 82247352.0},
+			Tags:   map[string]string{"peer-name": "AZDCLTEP1"},
+		},
+	}
+
+Definitions:
+	leaf - a single tag/field stored in *ProcessedResponse
+	branch - any *ProcessedResponse that isn't a leaf
+*/
+func ProcessResponse(jsonData *gabs.Container, fields map[string]string, tags map[string]string) ([]*ProcessedResponse, error) {
```

Comment:
i think we should be defensive and protect against nil jsonData ptr but returning error early